### PR TITLE
decompile unk_02006D98.s

### PR DESCRIPTION
**Pull request by remicalixte**
some low-level graphic stuff again.

I left a couple of non matching function that I couldn't figure out.
I struggled a lot with the data section this time, I don't quite understand how the compiler orders the symbols. It seems roughly based on their size but the types also affect the order. I also had to move 2 bytes from `UNK_020ECC78` to `UNK_020ECCE2` to change the size of `UNK_020ECC78` to a multiple of 4 otherwise the compiler kept adding extra padding. Obviously this is an ugly workaround that I'd like to remove but I'm currently out of ideas.

### CODE REVIEW

File: arm9/src/unk_02006D98.c [hunk 1]
```diff
@@ -0,0 +1,2855 @@
...
+#include "MI_memory.h"
+#include "NNS_g2d.h"
+#include "filesystem.h"
+#include "heap.h"
+#include "mod63_021DB450.h"
+#include "palette.h"
+#include "poketool/pokegra/otherpoke.naix"
+#include "registers.h"
+
+const u8 UNK_020ECCE0[] = {
```
Comment (PikalaxALT) [inline arm9/src/unk_02006D98.c:12]: Try to interpret these as the struct to which they are typecast. iirc structs are aligned to short on arm946e-s when the largest member data type is short or byte aligned.

**Review by PikalaxALT — CHANGES_REQUESTED** [review]

File: arm9/src/unk_02006D98.c (hunk 1 above)
Comment (remicalixte) [reply arm9/src/unk_02006D98.c]: I tried that before and it indeed fixes the padding but I couldn't get `UNK_020ECD4C` to be ordered properly. Just retried and now it works :man_shrugging:

Comment (remicalixte) [reply arm9/src/unk_02006D98.c]: I've noticed adding functions to a file sometimes changes the order of the symbols so maybe I just needed to finish decompiling first

Comment (PikalaxALT) [reply arm9/src/unk_02006D98.c]: Yeah, that's a consequence of compiler flag `-ipa file`.

File: arm9/src/unk_02006D98.c [hunk 2]
```diff
@@ -0,0 +1,2489 @@
...
+#include "NNS_g2d.h"
+#include "filesystem.h"
+#include "heap.h"
+#include "mod63_021DB450.h"
+#include "palette.h"
+#include "poketool/pokegra/otherpoke.naix"
+#include "registers.h"
+
+const struct UnkStruct_02006D98_3 UNK_020ECCE2[] = {
+{0x1B,0x0F},
```
Comment (red031000) [inline arm9/src/unk_02006D98.c:13]: indent these one?

File: include/unk_02006D98.h [hunk 3]
```diff
@@ -0,0 +1,179 @@
+#ifndef POKEDIAMOND_UNK_02006D98_H
+#define POKEDIAMOND_UNK_02006D98_H
+
+#include "global.h"
+#include "NNS_g2d.h"
+struct UnkStruct63_021DB49C
```
Comment (red031000) [inline include/unk_02006D98.h:6]: this doesn't need 63 in its name anymore

**Review by red031000 — CHANGES_REQUESTED** [review]

**Review by red031000 — CHANGES_REQUESTED** [review]
yeaaaah the rest of the style was good apart from what I pointed out, now it's not so good Kappa

**Comment by remicalixte** [comment]
Mmh yeah, I just ran clang-format on the whole file, I don't know why it did that. I won't be able to fix it for now since I don't have access to a computer

**Review by red031000 — DISMISSED** [review]

File: arm9/src/unk_02006D98.c [hunk 4]
```diff
@@ -0,0 +1,2488 @@
...
+    FreeToHeap(st4);
+    ptr->unk2E1 = 1;
+    ptr->unk2E2 = 1;
+
+    return ptr;
+}
+
+#ifdef NONMATCHING
+
+// mostly matching, only 2 register writes messed up
```
Comment (PikalaxALT) [inline arm9/src/unk_02006D98.c:355]: probably sdk inlines are to blame

File: arm9/src/unk_02006D98.c [hunk 5]
```diff
@@ -0,0 +1,2488 @@
...
+{
+    return param0->unk54_C == 1;
+}
+
+THUMB_FUNC void FUN_02007F34(struct UnkStruct_02006D98_2 *param0, s32 param1)
+{
+    param0->unk2E = (s16)((40 - param1) - (((40 - param1) * param0->unk36) >> 8));
+}
+
+#ifdef NONMATCHING
```
Comment (PikalaxALT) [inline arm9/src/unk_02006D98.c:1697]: what's the nonmatching behavior here?

File: arm9/src/unk_02006D98.c [hunk 6]
```diff
@@ -0,0 +1,2488 @@
...
+    param0[3] = param1->unk1;
+    ((struct UnkStruct_02006D98_sub **)param0)[4] = param1;
+
+    for (s32 i = 0; i < 10; i++)
+    {
+        param0[i + 4] = 0;
+    }
+}
+
+#ifdef NONMATCHING
```
Comment (PikalaxALT) [inline arm9/src/unk_02006D98.c:1871]: what's the nonmatching behavior here?

File: arm9/src/unk_02006D98.c (hunk 4 above)
Comment (remicalixte) [reply arm9/src/unk_02006D98.c:355]: What should we do in this case? Is it okay to copy them from the SDK? I see that some are already in the repo

File: arm9/src/unk_02006D98.c (hunk 6 above)
Comment (remicalixte) [reply arm9/src/unk_02006D98.c:1871]: Mainly register allocation issues. Both problematic functions are similar, maybe there is a macro/inline here too

File: arm9/src/unk_02006D98.c (hunk 4 above)
Comment (PikalaxALT) [reply arm9/src/unk_02006D98.c:355]: > Is it okay to copy them from the SDK?
Yes, just strip comments and sync the formatting/style to match the rest of the project.

**Comment by remicalixte** [comment]
Alright, no nonmatching left. inline funcs are so tricky...

**Comment by remicalixte** [comment]
Is there anything left I need to change? @PikalaxALT @red031000

File: arm9/src/unk_02006D98.c [hunk 7]
```diff
@@ -0,0 +1,1747 @@
...
+        else
+        {
+            shift = 0;
+        }
+
+        reg_G3_TEXPLTT_BASE = (param0->unk2A4 + st18 * 0x20) >> (4 - shift);
+
+        u32 r1 = param0->unk000[st18].unk28 << 0xc;
+        u32 r0 = (u32)((param0->unk000[st18].unk26 + param0->unk000[st18].unk42) << 0xc);
+        u32 r3 = (u32)((param0->unk000[st18].unk24 + param0->unk000[st18].unk40) << 0xc);
```
Comment (PikalaxALT) [inline arm9/src/unk_02006D98.c:410]: these are behaving like fx32 types and should be treated as such

File: arm9/src/unk_02006D98.c [hunk 8]
```diff
@@ -0,0 +1,1747 @@
...
+
+        reg_G3_MTX_TRANS = r3;
+        reg_G3_MTX_TRANS = r0;
+        reg_G3_MTX_TRANS = r1;
```
Comment (PikalaxALT) [inline arm9/src/unk_02006D98.c:414]: register access usually happens within the scope of an sdk inline

**Review by PikalaxALT — CHANGES_REQUESTED** [review]

File: arm9/src/unk_02006D98.c (hunk 8 above)
Comment (remicalixte) [reply arm9/src/unk_02006D98.c]: I added all the inlines used in this file, it matched without any effort

**Review by PikalaxALT — APPROVED** [review]

**Pull request merged**